File: n_queen_problem/n_queen_problem.py
"""
Following code is the implementation of the N-queen problem.
The solution is to place the queens on board such that
no queen is under attack by another queen.
This implementation computes all solutions for a board,
then prints the number of solutions. The number of solutions
follows the OEIS sequence https://oeis.org/A000170.
"""


def is_queen_under_attack(col, queens):
    """
    :param col: int
        Column of queen to test
    :param queens: list of tuple of int
        List of coordinates of other queens
    :return: boolean
        True if and only if queen in given position is under attack
    """

    left = right = col
    for coords in reversed(queens):
        left, right = left - 1, right + 1
        if coords[1] in (left, col, right):
            return True
    return False


def get_solutions(board_size, queens):
    """
    :param board_size: int
        Size of board to solve
    :param queens: int
        Number of queens to place in board
    :return: list of list of tuple of int
        Finds solutions of problem
    """

    smaller_solutions = n_queen_problem(board_size, queens - 1)
    solutions = []
    for solution in smaller_solutions:
        for column in range(1, board_size + 1):
            if not is_queen_under_attack(column, solution):
                solutions.append(solution + [(queens, column)])

    return solutions


def n_queen_problem(board_size, queens):
    """
    :param board_size: int
        Size of board to solve
    :param queens: int
        Number of queens to place in board
    :return: list of list of tuple of int
        Each list contains the coordinates of the queens
        that solve the problem
    """

    if queens < 1:  # there are no queens to place
        return [[]]  # 1 solution only: not place any queen

    if queens > board_size:  # more queens than the board can contain
        return []  # pigeonhole principle -> 0 solutions

    return get_solutions(board_size, queens)


def main():
    board_size = 10
    solutions = n_queen_problem(board_size, board_size)
    print(len(solutions))


if __name__ == '__main__':
    main()

File: n_queen_problem/n_queen_problem.go
// Given a chess board having N×N cells
// place N queens on the board in such a way that no queen attacks any other queen.

// There can be a number of possible solutions for a given length of board.
// This implementation prints only one valid solution, it can be extended to print all possible valid solutions.

package main

import "fmt"

func main() {
	n := 8 // size of chess board (n x n)

	// make board
	b := make([][]int, n)
	for i := range b {
		b[i] = make([]int, n)
	}

	if placeQueen(&b, n) {
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				fmt.Printf("%d ", b[i][j])
			}
			fmt.Print("\n")
		}
	} else {
		fmt.Println("Not possible")
	}
}

func isAttacked(b *[][]int, x, y, n int) bool {
	for i := 0; i < n; i++ {
		if (*b)[x][i] == 1 {
			return true
		}
		if (*b)[i][y] == 1 {
			return true
		}
		for j := 0; j < n; j++ {
			if (i-j == x-y) || (i+j == x+y) {
				if (*b)[i][j] == 1 {
					return true
				}
			}
		}
	}
	return false
}

func placeQueen(b *[][]int, n int) bool {
	if n == 0 {
		return true
	}
	for i := 0; i < len(*b); i++ {
		for j := 0; j < len(*b); j++ {
			if !(isAttacked(b, i, j, len(*b))) {
				(*b)[i][j] = 1
				if placeQueen(b, n-1) {
					return true
				}
				(*b)[i][j] = 0
			}
		}
	}
	return false
}
